stats/src/main/java/com/facebook/airlift/stats/QuantileDigest.java

```diff
@@ -1315,4 +1337,172 @@ public String toString()
 
         double middle(long lowerBound, long upperBound);
     }
+
+    public static class QuantileDigestIterator
```

Comment:
This iterator is not implementing the `Iterator` interface, which is quite unintuitive. Any reason behind that?

Reply:
There were a couple of options here and I didn't choose any of them:
1) I could have made this Iterator<Integer> for the sake of compatibility--perhaps this would represent the upper bound quantile.  But because the iteration represents multiple values, then you'd still need the other methods to query them.
2) I could have had a flat `QuantileDigestItem` sort of thing that holds the artifacts like upper/lower bound quantile, cumulative count, etc.
2.a) This could be one item per entry, and you'd incur the object allocation cost per iteration
2.b) This could be one single object which is mutated per iteration, with the gotcha that if you attempt to copy it into a list you'd have a list with all items being the same value.

Out of the above, if I had to pick, I'd pick 2.b, and just comment about the disclaimer.  But actually I just preferred to bake it into the type, where there is no gotcha related to object allocations or autoboxing, and it's relatively simple to just jump into the interface to understand its behavior (since an iterator is also a language-neutral data structure that doesn't necessarily have the same interface as the one included in `java.util`).

Reply:
@tdcmeehan : > It would also have an autoboxing cost on every .next().

Don't forgot there is `IntIterator` in `fastutil` (is it accessible in airlift though?)